File: src/main/kotlin/com/linqingying/cangjie/psi/CjBinaryExpressionWithTypeRHS.java
package com.linqingying.cangjie.psi;

import com.linqingying.cangjie.CjNodeTypes;
import com.intellij.lang.ASTNode;
import com.intellij.psi.PsiElement;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Objects;

public class CjBinaryExpressionWithTypeRHS  extends CjExpressionImpl implements CjOperationExpression{


    public CjBinaryExpressionWithTypeRHS(@NotNull ASTNode node) {
        super(node);
    }

    @Override
    public <R, D> R accept(@NotNull CjVisitor<R, D> visitor, @Nullable D data) {
        return visitor.visitBinaryWithTypeRHSExpression(this, data);
    }

    @NotNull
    public CjExpression getLeft() {
        CjExpression left = findChildByClass(CjExpression.class);
        assert left != null;
        return left;
    }

    @Nullable
    @IfNotParsed
    public CjTypeReference getRight() {
        ASTNode node = getOperationReference().getNode();
        while (node != null) {
            PsiElement psi = node.getPsi();
            if (psi instanceof CjTypeReference) {
                return (CjTypeReference) psi;
            }
            node = node.getTreeNext();
        }

        return null;
    }

    @Override
    @NotNull
    public CjSimpleNameExpression getOperationReference() {
        return Objects.requireNonNull(findChildByType(CjNodeTypes.OPERATION_REFERENCE));
    }
}

File: src/main/kotlin/com/linqingying/cangjie/psi/CjBinaryExpressionWithTypeRHS.kt
package com.linqingying.cangjie.psi

import com.linqingying.cangjie.CjNodeTypes
import com.intellij.lang.ASTNode
import java.util.*

class CjBinaryExpressionWithTypeRHS(node: ASTNode) : CjExpressionImpl(node), CjOperationExpression {
    override fun <R, D> accept(visitor: CjVisitor<R, D>, data: D?): R {
        return visitor.visitBinaryWithTypeRHSExpression(this, data)
    }

    val left: CjExpression
        get() {
            val left = checkNotNull(
                findChildByClass(
                    CjExpression::class.java
                )
            )
            return left
        }

    @get:IfNotParsed
    val right: CjTypeReference?
        get() {
            var node = operationReference.node
            while (node != null) {
                val psi = node.psi
                if (psi is CjTypeReference) {
                    return psi
                }
                node = node.treeNext
            }

            return null
        }

    override val operationReference : CjSimpleNameExpression get() =
           findChildByType(CjNodeTypes.OPERATION_REFERENCE)!!

}
